frappe/core/doctype/scheduled_job_type/scheduled_job_type.py

```diff
@@ -149,6 +154,11 @@ def update_scheduler_log(self, status):
 			self.scheduler_log.db_set("details", frappe.get_traceback())
 		if status == "Start":
 			self.db_set("last_execution", now_datetime(), update_modified=False)
+		if status == "Complete" and self.frequency in ("Hourly Long", "Daily Long"):
+			tick = get_scheduler_tick_interval()
+			jitter = randrange(0, 900, tick)
+			self.db_set("jitter", jitter, update_modified=False)
```

Comment:
Umm, this is one extra query on every run and we aren't doing anything else with the value. 

Why don't we directly add this random value in `get_next_execution`?

Reply:
If you mean something like this:

```python
def get_next_execution(self):
    # ...
    # EVERYTHING ELSE
    # ...
    
    random_value = randrange(0, 900, tick)  
    return next_execution + timedelta(seconds=random_value)
```

i.e without storing whether jitter was applied or not.

Then there is a certain probability that the job gets repeatedly jittered and executes outside the +900s window (or improbably, never, lol).

Of course you could then check the last executed time and see if it's outside the +900s window and execute it if it is, but then you kinda end up with the same issue that this PR is trying to solve.

Reply:
Croniter gives FIRST next execution based on last execution, so it's unlikely that it will ever get into a state where something never runs. 

https://github.com/frappe/frappe/blob/d715b0ad31147707592357ad1db22ffc78ebabfc/frappe/core/doctype/scheduled_job_type/scheduled_job_type.py#L113

2nd arg here is "start time", croniter doesn't care what's current time AFAIK.
For safety we can limit max jitter to 8-10 min for hourly jobs and 15 for daily.

Reply:
Responding for completeness.

---

I double checked and turns out, I'd missed the part where cron scheduling ignores difference between `last_scheduled` and `last_actual` (the next date is always w.r.t `last_scheduled + schedule_time_diff`).

So yeah, this: 

> unlikely that it will ever get into a state where something never runs.

is impossible, cause jitter is applied to scheduled amount which suffers no drift.

Worst case is of hourly jobs getting collected at the end of the 900s window. But the probability of that for a single job—assuming uniform distribution of `randrange` and `tick` of 60s—is `prob = 1/15; [prob := acc * (i/15) for i in range(2, 15)]` i.e. close to 0.

So yeah, just adding a random jitter amount is good enough.